internal/sftp/goclient.go

```diff
@@ -0,0 +1,84 @@
+package sftp
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"os"
+
+	"github.com/pkg/sftp"
+	"golang.org/x/crypto/ssh"
+)
+
+// GoClient implements the SFTP service using native Go SSH and SFTP packages.
+type GoClient struct {
+	cfg Config
+
+	ssh  *ssh.Client
+	sftp *sftp.Client
+}
+
+var _ Service = (*GoClient)(nil)
+
+// NewGoClient returns a new GoSFTP client with the given configuration.
+func NewGoClient(cfg Config) *GoClient {
+	cfg.SetDefaults()
+
+	return &GoClient{cfg: cfg}
+}
+
+// Upload writes the data from src to the remote file at dest and returns the
+// number of bytes written.  A new SFTP connection is opened before writing, and
+// closed when the upload is complete.
+func (c *GoClient) Upload(src io.Reader, dest string) (int64, error) {
+	if err := c.dial(); err != nil {
+		return 0, err
+	}
+	defer c.close()
+
+	// Note: Some SFTP servers don't support O_RDWR mode.
+	w, err := c.sftp.OpenFile(dest, (os.O_WRONLY | os.O_CREATE | os.O_TRUNC))
+	if err != nil {
+		return 0, fmt.Errorf("SFTP: couldn't create remote file %q: %w", dest, err)
+	}
+
+	bytes, err := io.Copy(w, src)
+	if err != nil {
+		return 0, fmt.Errorf("SFTP: failed to write to %q: %w", dest, err)
+	}
+
+	return bytes, nil
+}
```

Comment:
Not critical, but one issue with this implementation is that we can't pass a cancellation signal via context, e.g. if the Temporal activity is cancelled, etc. Ideally you'd accept a `ctx context.Context` as the first parameter. `io.Copy` can't use context out of the box but https://pkg.go.dev/github.com/dolmen-go/contextio provides context-aware wrappers for readers that you could use. We used it successfully in sdps_preprocessing (e.g. `upload_dip.go` if you want to take a look).
Maybe a future improvement or a small change you can make as part of this pull request if time permits!
Related: https://pace.dev/blog/2020/02/03/context-aware-ioreader-for-golang-by-mat-ryer.html.

Reply:
Yeah, that's a good idea considering these uploads could take a long time. :+1: